lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_firmware.py

```diff
@@ -0,0 +1,209 @@
+       reboot:     True
+       job_wait:   True
+       catalog_file_name:  "Catalog.xml"
+"""
+
+RETURN = """
+---
+msg:
+    description: Updates firmware from a repository on a network share (CIFS, NFS).
+    returned: success
+    type: string
+    sample: "{
+        'CompletionTime': '2018-01-24T00:47:33',
+        'Id': 'JID_XXXXXXXXXXXX',
+        'JobState': 'Completed',
+        'JobType': 'ImportConfiguration',
+        'Message': 'Successfully imported and applied Server Configuration Profile.',
+        'MessageId': 'SYSXXX',
+        'Name': 'Import Configuration',
+        'StartTime': 'TIME_NOW',
+        'Status': 'Success',
+    }"
+"""
+
+
+from ansible.module_utils.remote_management.dellemc.dellemc_idrac import iDRACConnection
+from ansible.module_utils.basic import AnsibleModule
+try:
+    from omsdk.sdkcreds import UserCredentials
+    from omsdk.sdkfile import FileOnShare
+    HAS_OMSDK = True
+except ImportError:
+    HAS_OMSDK = False
+
+
+def _validate_catalog_file(catalog_file_name):
+    normilized_file_name = catalog_file_name.lower()
+    if not normilized_file_name:
+        raise ValueError('catalog_file_name should be a non-empty string.')
+    elif not normilized_file_name.endswith("xml"):
+        raise ValueError('catalog_file_name should be an XML file.')
+
+
+def update_firmware(idrac, module):
+    """Update firmware from a network share and return the job details."""
+    msg = {}
+    msg['changed'] = False
+    msg['update_status'] = {}
+
+    try:
+        _validate_catalog_file(module.params['catalog_file_name'])
+    except ValueError as e:
+        module.fail_json(msg=str(e))
+
+    try:
+        upd_share = FileOnShare(remote=module.params['share_name'] + "/" + module.params['catalog_file_name'],
+                                mount_point=module.params['share_mnt'],
+                                isFolder=False,
+                                creds=UserCredentials(
+                                    module.params['share_user'],
+                                    module.params['share_pwd'])
+                                )
+
+        idrac.use_redfish = True
+        if '12' in idrac.ServerGeneration or '13' in idrac.ServerGeneration:
+            idrac.use_redfish = False
+
+        apply_update = True
+        msg['update_status'] = idrac.update_mgr.update_from_repo(upd_share,
+                                                                 apply_update,
+                                                                 module.params['reboot'],
+                                                                 module.params['job_wait'])
+    except Exception as e:
+        module.fail_json(msg=str(e))
+
+    if "Status" in msg['update_status']:
+        if msg['update_status']['Status'] == "Success":
+            if module.params['job_wait'] is True:
+                msg['changed'] = True
+        else:
+            module.fail_json(msg=msg['update_status'])
```

Comment:
`update_status` should be a separate key in return structure, so you can still add some msg text string:
```suggestion
            module.fail_json(msg='Firmware update failed', update_status=msg['update_status'])
```

Reply:
updated the required changes for this `fail_json `